### PR TITLE
Fix constraint simplification for inferred predicates

### PR DESCRIPTION
**Pull request by asubramanya**
When we encounter a CExpression that cannot be converted back to a
CConstraintInterval, just return the original constraint without
simplifying it - instead of asserting.

Co-authored-by: Abhijit Subramanya <asubramanya@pivotal.io>
Co-authored-by: Hans Zeller <hzeller@pivotal.io>

### CODE REVIEW

File: libgpopt/src/base/CConstraint.cpp [hunk 1]
```diff
@@ -780,6 +780,14 @@ CConstraint::PdrgpcnstrDeduplicate
...
 		CConstraint *pcnstrNew = CConstraintInterval::PciIntervalFromScalarExpr(mp, pexpr, colref);
+		if (NULL == pcnstrNew)
+		{
+			// We ran into a type conflict that prevents us from using this method to simplify the constraint.
+			// Give up and return the un-flattened constraint.
+			pcnstrChild->AddRef();
+			pcnstrNew = pcnstrChild;
+		}
+
 		GPOS_ASSERT(NULL != pcnstrNew);
```
Comment (vraghavan78) [inline libgpopt/src/base/CConstraint.cpp]: do we need this assert then?

Comment (asubramanya) [reply libgpopt/src/base/CConstraint.cpp]: Yeah. We can remove this.

**Comment by asubramanya** [comment]
@hardikar We tried the query you mentioned, but did not hit our code change at all. Its true for both the string and the integer versions of the query. We will add some more comment in the code.

**Comment by asubramanya** [comment]
@chrishajas We will investigate the reason for not being able to convert the expression into `CConstraintInterval` separately. For now this fix should be fine since there is no reason to crash in this case.

**Pull request merged**